java/ql/lib/semmle/code/java/os/OSCheck.qll

```diff
@@ -0,0 +1,161 @@
+ */
+abstract class IsSpecificUnixVariant extends Guard { }
+
+/**
+ * Holds when `ma` compares the current OS against the string constant `osString`.
+ */
+private predicate isOsFromSystemProp(MethodAccess ma, string osString) {
+  TaintTracking::localExprTaint(getSystemProperty("os.name"), ma.getQualifier()) and // Call from System.getProperty (or equivalent) to some partial match method
+  exists(StringPartialMatchMethod m, CompileTimeConstantExpr matchedStringConstant |
+    m = ma.getMethod() and
+    matchedStringConstant.getStringValue().toLowerCase() = osString
+  |
+    DataFlow::localExprFlow(matchedStringConstant, ma.getArgument(m.getMatchParameterIndex()))
+  )
+}
+
+private class IsWindowsFromSystemProp extends IsWindowsGuard instanceof MethodAccess {
+  IsWindowsFromSystemProp() { isOsFromSystemProp(this, any(string s | s.regexpMatch("windows?"))) }
+}
+
+/**
+ * Holds when the Guard is an equality check between the system property with the name `propertyName`
+ * and the string or char constant `compareToLiteral`, and the branch evaluates to `branch`.
+ */
+private Guard isOsFromSystemPropertyEqualityCheck(
+  string propertyName, string compareToLiteral, boolean branch
+) {
+  result
+      .isEquality(getSystemProperty(propertyName),
+        any(Literal literal |
+          (literal instanceof CharacterLiteral or literal instanceof StringLiteral) and
+          literal.getValue() = compareToLiteral
+        ), branch)
+}
+
+private class IsWindowsFromPathSeparator extends IsWindowsGuard {
+  IsWindowsFromPathSeparator() {
+    this = isOsFromSystemPropertyEqualityCheck("path.separator", ";", true) or
+    this = isOsFromSystemPropertyEqualityCheck("path.separator", ":", false)
+  }
+}
+
+private class IsWindowsFromFileSeparator extends IsWindowsGuard {
+  IsWindowsFromFileSeparator() {
+    this = isOsFromSystemPropertyEqualityCheck("file.separator", "\\", true) or
+    this = isOsFromSystemPropertyEqualityCheck("file.separator", "/", false)
+  }
+}
+
+private class IsUnixFromPathSeparator extends IsUnixGuard {
+  IsUnixFromPathSeparator() {
+    this = isOsFromSystemPropertyEqualityCheck("path.separator", ":", true) or
+    this = isOsFromSystemPropertyEqualityCheck("path.separator", ";", false)
+  }
+}
+
+private class IsUnixFromFileSeparator extends IsUnixGuard {
+  IsUnixFromFileSeparator() {
+    this = isOsFromSystemPropertyEqualityCheck("file.separator", "/", true) or
+    this = isOsFromSystemPropertyEqualityCheck("file.separator", "\\", false)
+  }
+}
+
+private class IsUnixFromSystemProp extends IsSpecificUnixVariant instanceof MethodAccess {
+  IsUnixFromSystemProp() {
+    isOsFromSystemProp(this, any(string s | s.regexpMatch(["mac.*", "linux.*"])))
+  }
+}
+
+bindingset[fieldNamePattern]
+private predicate isOsFromApacheCommons(FieldAccess fa, string fieldNamePattern) {
+  exists(Field f | f = fa.getField() |
+    f.getDeclaringType() instanceof TypeApacheSystemUtils and
+    f.getName().matches(fieldNamePattern)
```

Comment:
I guess it is unlikely that this will give spurious matches, but do note that `_` is a wildcard for the `.matches` predicate, so all of the underscores in the names below could be matched with anything.